src/vm/olevariant.h

```diff
@@ -324,9 +326,16 @@ struct VariantData
     }
 
 private:
-    Object*     m_or;
+#ifdef _WIN64
+    Object*     m_objref;
+    INT64       m_data;
+    INT32       m_flags;
+    INT32       m_padding;
+#else
     INT64       m_data;
```

Comment:
The order of these fields in the `#else` is wrong. I would suggest we simply add the `m_padding` field in a `#ifdef _WIN64` or am I missing something else here?

Reply:
The order is right. The type loader reorders the fields for non-blittable types and this reordering differs for 32-bit vs. 64-bit.

Reply:
>The type loader reorders the fields for non-blittable types and this reordering differs for 32-bit vs. 64-bit.

Can we comment that here. Do we have a link somewhere so I can educate myself?

Reply:
A link to the general area in the code base would be enough too.

Reply:
The type loader is trying to maximize the packing and that's why it reorders the fields when it is allowed to do so.

The type loader lives in https://github.com/dotnet/coreclr/blob/master/src/vm/methodtablebuilder.cpp . There are number of different methods involved in the layout - PlaceInstanceFields is probably the more interesting ones.

There are more types used with manually-managed code that have similar difference. E.g.  https://github.com/dotnet/coreclr/blob/master/src/vm/object.h#L1627

Reply:
>The type loader is trying to maximize the packing and that's why it reorders the fields when it is allowed to do so.

I figured that would be the rational. Wasn't aware reordering was automatically done by the CLR. Is there publicly documented? Thanks for the source links.

Reply:
We do document that non-blittable types have different managed and unmanaged layout: https://docs.microsoft.com/en-us/dotnet/framework/interop/blittable-and-non-blittable-types

We do not document the specific optimizations that we apply to the layout, but you can find approximate descriptions of them on stackoverflow and similar places.